presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,541 @@
+        {
+            checkState(isEmpty(), "bucket %s not empty, points to %s", bucketId, buckets.get(bucketId));
+
+            // we've already computed the value hash for only the value only; ValueStore will save it for future use
+            int nextValuePosition = valueStore.getPosition(type, block, position, valueHash);
+            // set value pointer to hash map of values
+            valuePositions.set(nodePointerToUse, nextValuePosition);
+            // save hashes for future rehashing
+            valueHashList.set(nodePointerToUse, valueHash);
+            // set pointer to node for this bucket
+            buckets.set(bucketId, nodePointerToUse);
+            // save data for this node
+            counts.set(nodePointerToUse, count);
+            // used for doing value comparisons on hash collisions
+            groupIds.set(nodePointerToUse, groupId);
+            // we only ever store ints as values; we need long as an index
+            int currentHead = (int) headPointers.get(groupId);
+            // maintain linked list of nodes in this group (insert at head)
+            headPointers.set(groupId, nodePointerToUse);
+            nextPointers.set(nodePointerToUse, currentHead);
+        }
+    }
+
+    private class BucketNodeFactory
+    {
+        /**
+         * invariant: "points" to a virtual node of [key, count] for the histogram and includes any indirection calcs. Makes not guarantees if the node is empty or not
+         * (use isEmpty())
+         *
+         * @param currentBuckets
+         * @param i
+         * @param groupId
+         * @param block
+         * @param position
+         */
+        private BucketDataNode createBucketDataNode(long groupId, Block block, int position)
+        {
+            long valueHash = TypeUtils.hashPosition(type, block, position);
+            long valueAndGroupHash = getHash(groupId, valueHash);
+            int bucketId = (int) (valueAndGroupHash & mask);
+            int nodePointer;
+            int probeCount = 1;
+            int originalBucketId = bucketId;
+            // look for an empty slot or a slot containing this key
+            boolean isEmpty;
+            while (true) {
+                nodePointer = buckets.get(bucketId);
+                // empty
+                if (nodePointer == EMPTY_BUCKET) {
+                    isEmpty = true;
+                    break;
+                }
+                int valuePosition = valuePositions.get(nodePointer);
+                // value match
+                if (groupAndValueMatches(groupId, block, position, nodePointer, valuePosition)) {
+                    isEmpty = false;
+                    break;
+                }
+                int probe = nextProbe(probeCount);
+                bucketId = nextBucketId(originalBucketId, mask, probe);
+                probeCount++;
+            }
+
+            int nodePointerToUse = isEmpty ? nextNodePointer : nodePointer;
```

Comment:
If you inline this into each branch within the loop above, the code would simplify a bit:

```java
while (true) {
    nodePointer = buckets.get(bucketId);
    // empty
    if (nodePointer == EMPTY_BUCKET) {
        return new BucketDataNode(bucketId, new ValueNode(nextNodePointer), valueHash, nextNodePointer, true);
    }
    int valuePosition = valuePositions.get(nodePointer);
    // value match
    if (groupAndValueMatches(groupId, block, position, nodePointer, valuePosition)) {
        return new BucketDataNode(bucketId, new ValueNode(nodePointer), valueHash, nodePointer, false);
    }
    int probe = nextProbe(probeCount);
    bucketId = nextBucketId(originalBucketId, mask, probe);
    probeCount++;
}
```

Reply:
similar to valueStore ?

ok will do

Reply:
btw 
this is a pattern i follow: single return point from functions. i guess it's arguably more complex, but it makes reasoning about exit points easy. it does make reading the state of vars and if the final exit gets much more complex than this, it can be a net loss

i've made the change, but still lean towards
1. code body of loops, if-else, etc collects data, possibly constructs objects for return and assigns to local variable
2. final statement returns object, or if is necessary, constructs and returns

not sure if there are any jvm perf impls on loops either ways? (asking?)

Reply:
done